h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java

```diff
@@ -48,19 +48,29 @@ public void map(Chunk[] chk, NewChunk[] newGamCols) {
     for (int rowIndex = 0; rowIndex < chunkRows; rowIndex++) {
```

Comment:
The chunks might have different length !

Reply:
do you mean `chk[i].len() != chk[j].len()` for some i,j? that should not be possible

Reply:
Exactly.

```
    for (int i = 0; i < chk.length; i++) {
      if(chk[0].len() != chk[i].len()){
        throw new IllegalArgumentException(String.format("Chunks have different length. %s vs %s", chk[0].len(), chk[i].len()));
      }
    }
```

is actually triggered (even locally).

Reply:
Here is a full patch to easily apply:

```patch
Index: h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java	(revision c40a93bb070fb9c4d55c95858ec80d13e956f3f8)
+++ h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java	(date 1600675633440)
@@ -39,6 +39,11 @@
 
   @Override
   public void map(Chunk[] chk, NewChunk[] newGamCols) {
+    for (int i = 0; i < chk.length; i++) {
+      if(chk[0].len() != chk[i].len()){
+        throw new IllegalArgumentException(String.format("Chunks have different length. %s vs %s", chk[0].len(), chk[i].len()));
+      }
+    }
     _maxAbsRowSum = new double[_gamX.vec(0).nChunks()];
     int cIndex = chk[0].cidx();
     _maxAbsRowSum[cIndex] = Double.NEGATIVE_INFINITY;
```

Reply:
Next step is obvious, look at the input and see how the chunks are created and whether the layout is the same. At first sight, it should NOT be possible as those are added to a new Frame and checks are NOT disabled, but that `Frame` thing has like a million of constructors, so maybe I overlooked it.

Let the digging continue.

Reply:
I dont think its related to anyhing but I find it interesting that every map() call overwrites the array `_maxAbsRowSum = new double[_gamX.vec(0).nChunks()];`

Reply:
That is just a bad design I believe. But thank you, every clue, no matter how small is important. I investigated this one as well. The Frame itself can be omitted from serialization and this value can just be initialized properly where it should be (inside the constructor). This has no effect on the error (see diff below to quickly apply if you feel like it)

```patch
Index: h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java	(revision c40a93bb070fb9c4d55c95858ec80d13e956f3f8)
+++ h2o-algos/src/main/java/hex/gam/MatrixFrameUtils/GenerateGamMatrixOneColumn.java	(date 1600676523134)
@@ -19,7 +19,6 @@
   int _splineType;
   public int _numKnots;      // number of knots
   public double[][] _bInvD;  // store inv(B)*D
-  Frame _gamX;
   double[] _u; // store transpose(X)*1, sum across rows per column
   public double[][] _ZTransp;  // store Z matrix transpose
   public double[][] _penaltyMat;  // store penalty matrix
@@ -33,13 +32,12 @@
      CubicRegressionSplines crSplines = new CubicRegressionSplines(numKnots, knots);
     _bInvD = crSplines.gen_BIndvD(crSplines._hj);
     _penaltyMat = crSplines.gen_penalty_matrix(crSplines._hj, _bInvD);
-    _gamX = gamx;
+    _maxAbsRowSum = new double[gamx.vec(0).nChunks()];
     _knots = knots;
   }
 
   @Override
   public void map(Chunk[] chk, NewChunk[] newGamCols) {
-    _maxAbsRowSum = new double[_gamX.vec(0).nChunks()];
     int cIndex = chk[0].cidx();
     _maxAbsRowSum[cIndex] = Double.NEGATIVE_INFINITY;
     int chunkRows = chk[0].len(); // number of rows in chunk

```

Reply:
Both vectors in the `predictVec` Frame (that naming !) report the same length and the same `espc`:

```java
            GenerateGamMatrixOneColumn genOneGamCol = new GenerateGamMatrixOneColumn(splineType, numKnots, 
                    _knots[frameIndex], predictVec).doAll(numKnots, Vec.T_NUM, predictVec);
```

On this level, everything seems to be just fine.

Reply:
Also, one more observation. Notice the whole MRTask is done in parallel in a batch of `RecursiveAction` - not sure why we need recursive action for that. Yet I've tried to make this sequential to quickly rule out a problem with concurrent modification on that side, to no avail. Problem still remains.

Reply:
Oh wow,

```java
        final Frame predictVec = new Frame();
        predictVec.add(_parms._gam_columns[index], _parms._train.get().vec(_parms._gam_columns[index]));
        predictVec.add("weights_column", weights_column);
```

This seems to produce even chunks and proper layout. Wow. Let's try that.

Reply:
Hooray, Fixed ! http://mr-0xc1:8080/blue/organizations/jenkins/h2o-3-pipeline/detail/PR-4869/44/tests/